src/language-markdown/printer-markdown.js

```diff
@@ -52,6 +56,43 @@ const SIBLING_NODE_TYPES = new Set([
   "definition",
   "footnoteDefinition",
 ]);
+// https://en.wikipedia.org/wiki/Line_breaking_rules_in_East_Asian_languages
+/**
+ * The set of characters that must not immediately precede a line break
+ *
+ * e.g. `"（"`
+ *
+ * - Bad:  `"檜原村（\nひのはらむら）"`
+ * - Good: `"檜原村\n（ひのはらむら）"` or ``"檜原村（ひ\nのはらむら）"`
+ */
+const noBreakAfterCharacterSet = new Set(
+  "$(£¥·'\"〈《「『【〔〖〝﹙﹛＄（［｛￡￥[{‵︴︵︷︹︻︽︿﹁﹃﹏〘｟«"
+);
+/**
+ * The set of characters that must not immediately follow a line break
+ *
+ * e.g. `"）"`
+ *
+ * - Bad:  `"檜原村（ひのはらむら\n）以外には、"`
+ * - Good: `"檜原村（ひのはらむ\nら）以外には、"` or `"檜原村（ひのはらむら）\n以外には、"`
+ */
+const noBreakBeforeCharacterSet = new Set(
+  "!%),.:;?]}¢°·'\"†‡›℃∶、。〃〆〕〗〞﹚﹜！＂％＇），．：；？］｝～–—•〉》」︰︱︲︳﹐﹑﹒﹓﹔﹕﹖﹘︶︸︺︼︾﹀﹂﹗｜､』】〙〟｠»ヽヾーァィゥェォッャュョヮヵヶぁぃぅぇぉっゃゅょゎゕゖㇰㇱㇲㇳㇴㇵㇶㇷㇸㇹㇺㇻㇼㇽㇾㇿ々〻‐゠〜～‼⁇⁈⁉・"
+);
+/**
+ * The set of characters whose surrounding newline may be converted to Space
+ *
+ * - ASCII punctuation marks
+ */
+const lineBreakBetweenTheseAndCJKConvertToSpaceSymbolSet = new Set(
```

Comment:
This name is very long, but I have no idea how to make it shorter. Please change "convert" to "converts" and "symbol" to "character".

Reply:
We can remove `Set`, `...Characters`

Reply:
I regret `SymbolSet` is a little shorter than `Characters`...

Reply:
@thorn0 @fisker How about `charactersToConvertNewlineBetweenCJAndToSpace` (50→45)?

Reply:
The old name was better. Let's keep it.

Reply:
I propose to remove `CharacterSet` from all the names. We don't name other variables like that (e.g. `printWidth`, not `printWidthNumber`). A type can be added to the name to distinguish otherwise similar variables (e.g. `punctuationPattern` and `punctuationRegex`), but it's not the case for these character set variables.

Reply:
@thorn0 Done for `noBreak{Before,After}(CharactersSet)`